### PR TITLE
Add MAX_FOUND_METRICS parameter to prevent from too large metrics

### PR DESCRIPTION
**Pull request by Thib17**
Enable limiting metric list retuned by the StandardFinder.
Do not limit as default.

### CODE REVIEW

File: webapp/graphite/finders/standard.py [hunk 1]
```diff
@@ -67,8 +75,12 @@ def find_nodes(self, query):
               for datasource_name in RRDReader.get_datasources(absolute_path):
                 if match_entries([datasource_name], datasource_pattern):
                   reader = RRDReader(absolute_path, datasource_name)
+                  found_leafs += 1
                   yield LeafNode(metric_path + "." + datasource_name, reader)
 
+    if truncated:
+      log.exception('Truncated metrics finding due to MAX_FOUND_METRICS limit (%d) reached for "%s"' % (settings.MAX_FOUND_METRICS, clean_pattern))
```
Comment (iksaif) [inline webapp/graphite/finders/standard.py]: I think log.exception should be used only in an exception handler.

File: webapp/graphite/settings.py [hunk 2]
```diff
@@ -85,6 +85,7 @@
 LOG_ROTATION = True
 LOG_ROTATION_COUNT = 1
 MAX_FETCH_RETRIES = 2
+MAX_FOUND_METRICS = 0
```
Comment (iksaif) [inline webapp/graphite/settings.py]: I propose:
```
# Fail if more than X metrics are returned
METRIC_FIND_FAILURE_THRESHOLD = 0
# Truncate if more than X metrics are returned
METRIC_FIND_TRUNCATE_THRESHOLD = 0
# Print a warning if more than X metrics are returned
METRICS_FIND_WARNING_THRESHOLD = 0
```

File: webapp/graphite/finders/standard.py [hunk 3]
```diff
@@ -27,8 +27,14 @@ def find_nodes(self, query):
     clean_pattern = query.pattern.replace('\\', '')
     pattern_parts = clean_pattern.split('.')
 
+    found_leafs = 0
+    truncated = False
     for root_dir in self.directories:
       for absolute_path in self._find_paths(root_dir, pattern_parts):
+        if settings.MAX_FOUND_METRICS and found_leafs >= settings.MAX_FOUND_METRICS:
```
Comment (iksaif) [inline webapp/graphite/finders/standard.py]: Set it to +inf, this way you the code will be easier to read.

File: webapp/graphite/settings.py (hunk 2 above)
Comment (iksaif) [reply webapp/graphite/settings.py]: Note: you may want to implement METRIC_FIND_FAILURE_THRESHOLD and METRICS_FIND_WARNING_THRESHOLD in the place that calls all the finders.

**Review by iksaif — CHANGES_REQUESTED** [review]

File: webapp/graphite/finders/standard.py (hunk 3 above)
Comment (Thib17) [reply webapp/graphite/finders/standard.py]: Done

File: webapp/graphite/finders/standard.py (hunk 1 above)
Comment (Thib17) [reply webapp/graphite/finders/standard.py]: Done

File: webapp/graphite/settings.py (hunk 2 above)
Comment (Thib17) [reply webapp/graphite/settings.py]: Done

File: webapp/graphite/storage.py [hunk 4]
```diff
@@ -41,9 +41,16 @@ def __init__(self, finders=None, hosts=None):
   def find(self, pattern, startTime=None, endTime=None, local=False, headers=None):
     query = FindQuery(pattern, startTime, endTime, local)
 
+    matched_leafs = 0
     for match in self.find_all(query, headers):
+      if type(match) == LeafNode:
+          matched_leafs += 1
       yield match
 
+    if matched_leafs >= settings.METRICS_FIND_WARNING_THRESHOLD:
```
Comment (iksaif) [inline webapp/graphite/storage.py]: I find it easier to read when intermediate variables are used
```
warn_threshold =  settings.METRICS_FIND_WARNING_THRESHOLD
if matched_leafs > warn_treshold:
 log.rendering(..., warn_threhsold)
```

File: webapp/graphite/settings.py [hunk 5]
```diff
@@ -86,6 +86,11 @@
 LOG_ROTATION_COUNT = 1
 MAX_FETCH_RETRIES = 2
 
+METRICS_FIND_WARNING_THRESHOLD = float('Inf') # Print a warning if more than X metrics are returned
```
Comment (iksaif) [inline webapp/graphite/settings.py]: You may want to explain more why we need these settings.

File: webapp/graphite/finders/standard.py [hunk 6]
```diff
@@ -27,6 +27,7 @@ def find_nodes(self, query):
     clean_pattern = query.pattern.replace('\\', '')
     pattern_parts = clean_pattern.split('.')
 
+    found_leafs, truncated_leafs = 0, 0
```
Comment (iksaif) [inline webapp/graphite/finders/standard.py]: I think could you put both thresholds in intermediary variables, that would make the code easier to read:
truncate_threshold = setings.BLABLA..

File: webapp/graphite/finders/standard.py [hunk 7]
```diff
@@ -52,10 +53,18 @@ def find_nodes(self, query):
 
         elif isfile(absolute_path):
           if absolute_path.endswith('.wsp') and WhisperReader.supported:
+            if found_leafs >= settings.METRICS_FIND_WHISPER_TRUNCATE_THRESHOLD:
```
Comment (iksaif) [inline webapp/graphite/finders/standard.py]: you can yield later to factorize the code:

node = None
if whsiper:
  node = LeafNode(...)
elif gzip:
  node = LeafNode(...)

if node is not None:
   threshold checks
   yield node

File: webapp/graphite/finders/standard.py [hunk 8]
```diff
@@ -66,9 +75,21 @@ def find_nodes(self, query):
...
+                    continue
+                  found_leafs += 1
                   reader = RRDReader(absolute_path, datasource_name)
                   yield LeafNode(metric_path + "." + datasource_name, reader)
 
+    total_leafs = found_leafs + truncated_leafs
+    if total_leafs >= settings.METRICS_FIND_WHISPER_FAILURE_THRESHOLD:
+      raise Exception("Failed metrics find due to METRICS_FIND_WHISPER_FAILURE_THRESHOLD reached (%d > %d) for '%s'" %
+                      (total_leafs, settings.METRICS_FIND_WHISPER_FAILURE_THRESHOLD, clean_pattern))
+    elif truncated_leafs > 0:
```
Comment (iksaif) [inline webapp/graphite/finders/standard.py]: don't you want to exit the loop here ?

**Review by iksaif — CHANGES_REQUESTED** [review]

File: webapp/graphite/storage.py (hunk 4 above)
Comment (Thib17) [reply webapp/graphite/storage.py]: Done

File: webapp/graphite/settings.py (hunk 5 above)
Comment (Thib17) [reply webapp/graphite/settings.py]: Done

File: webapp/graphite/finders/standard.py (hunk 6 above)
Comment (Thib17) [reply webapp/graphite/finders/standard.py]: As discussed, refocussed on global threshold for now. Will work later on whisper finder

File: webapp/graphite/finders/standard.py (hunk 7 above)
Comment (Thib17) [reply webapp/graphite/finders/standard.py]: As discussed, refocussed on global threshold for now. Will work later on whisper finder

File: webapp/graphite/finders/standard.py (hunk 8 above)
Comment (Thib17) [reply webapp/graphite/finders/standard.py]: As discussed, refocussed on global threshold for now. Will work later on whisper finder

File: setup.py [hunk 9]
```diff
@@ -69,7 +69,7 @@
 try:
     setup(
       name='graphite-web',
-      version='1.1.0',
+      version='1.1.0-criteo1',
```
Comment (iksaif) [inline setup.py]: can you rebase and remove this commit ? :)

File: webapp/graphite/storage.py [hunk 10]
```diff
@@ -41,9 +41,21 @@ def __init__(self, finders=None, hosts=None):
...
+    matched_leafs = 0
     for match in self.find_all(query, headers):
+      if type(match) == LeafNode:
+          matched_leafs += 1
+      if matched_leafs > fail_threshold:
+        raise Exception("Failed metrics find due to METRICS_FIND_FAILURE_THRESHOLD reached (%d) for '%s'" % (fail_threshold, clean_pattern))
       yield match
 
+    if matched_leafs > warn_threshold:
+      log.rendering("warning : Large metrics find reaching METRICS_FIND_WARNING_THRESHOLD (%d > %d) for '%s'" %
```
Comment (iksaif) [inline webapp/graphite/storage.py]: "Query %s on %s yields more than %d results (warning threshold is %d)"

File: webapp/graphite/storage.py [hunk 11]
```diff
@@ -41,9 +41,21 @@ def __init__(self, finders=None, hosts=None):
...
 
+    warn_threshold = settings.METRICS_FIND_WARNING_THRESHOLD
+    fail_threshold = settings.METRICS_FIND_FAILURE_THRESHOLD
+
+    matched_leafs = 0
     for match in self.find_all(query, headers):
+      if type(match) == LeafNode:
+          matched_leafs += 1
+      if matched_leafs > fail_threshold:
+        raise Exception("Failed metrics find due to METRICS_FIND_FAILURE_THRESHOLD reached (%d) for '%s'" % (fail_threshold, clean_pattern))
```
Comment (iksaif) [inline webapp/graphite/storage.py]: "Query %s on %s yields more than %d results (failure threshold is %d)"

File: setup.py (hunk 9 above)
Comment (Thib17) [reply setup.py]: Done

File: webapp/graphite/storage.py (hunk 10 above)
Comment (Thib17) [reply webapp/graphite/storage.py]: Done

File: webapp/graphite/storage.py (hunk 11 above)
Comment (Thib17) [reply webapp/graphite/storage.py]: Done

**Review by iksaif — APPROVED** [review]

**Pull request merged**